# HG changeset patch
# User John Donoghue
# Date 1534354192 14400
#      Wed Aug 15 13:29:52 2018 -0400
# Branch stable
# Node ID 052a0d5dfd2cea03177ad2d0ba82719a5606754c
# Parent  8800e167c66514a0d113bbeebc52234e67968963
Reenable drag and drop of editor text (Bug #54495)

* libgui/src/m-editor/octave-qscintilla.h,
  libgui/src/m-editor/octave-qscintilla.cc
  (dragEnterEvent): new function to reject url mime drags

* libgui/src/m-editor/file-editor-tab.cc:
  (file_editor_tab): remove disable for editor accept drops

diff --git a/libgui/src/m-editor/file-editor-tab.cc b/libgui/src/m-editor/file-editor-tab.cc
--- a/libgui/src/m-editor/file-editor-tab.cc
+++ b/libgui/src/m-editor/file-editor-tab.cc
@@ -107,19 +107,16 @@ namespace octave
     _edit_area = new octave_qscintilla (this);
     _line = 0;
     _col  = 0;
 
     _bp_lines.clear ();      // start with empty lists of breakpoints
     _bp_conditions.clear ();
     m_bp_restore_count = 0;
 
-    // disable editor drag & drop so parent can handle
-    _edit_area->setAcceptDrops (false);
-
     connect (_edit_area, SIGNAL (cursorPositionChanged (int, int)),
              this, SLOT (handle_cursor_moved (int,int)));
 
     connect (_edit_area, SIGNAL (SCN_CHARADDED (int)),
              this, SLOT (handle_char_added (int)));
 
     connect (_edit_area, SIGNAL (SCN_DOUBLECLICK (int, int, int)),
              this, SLOT (handle_double_click (int, int, int)));
diff --git a/libgui/src/m-editor/octave-qscintilla.cc b/libgui/src/m-editor/octave-qscintilla.cc
--- a/libgui/src/m-editor/octave-qscintilla.cc
+++ b/libgui/src/m-editor/octave-qscintilla.cc
@@ -45,16 +45,18 @@ along with Octave; see the file COPYING.
 
 #include <Qsci/qscicommandset.h>
 
 #include <QKeySequence>
 #include <QShortcut>
 #include <QToolTip>
 #include <QVBoxLayout>
 
+#include <QMimeData>
+
 // FIXME: hardwired marker numbers?
 #include "marker.h"
 
 #include "octave-qscintilla.h"
 #include "file-editor-tab.h"
 #include "shortcut-manager.h"
 #include "resource-manager.h"
 #include "octave-settings.h"
@@ -957,11 +959,26 @@ namespace octave
       {
         if (first_word == "unwind_protect")
           first_word = '_' + first_word;
         next_line = "end" + first_word + "\n";
       }
 
     insertAt (QString (start, ' ') + next_line, linenr + 2, 0);
   }
+
+  void octave_qscintilla::dragEnterEvent (QDragEnterEvent *e)
+  {
+    // if is not dragging a url, pass to qscintilla to handle,
+    // otherwise ignore it so that it will be handled by
+    // the parent
+    if (!e->mimeData ()->hasUrls ())
+      {
+        QsciScintilla::dragEnterEvent (e);
+      }
+    else
+      {
+        e->ignore();
+      }
+  }
 }
 
 #endif
diff --git a/libgui/src/m-editor/octave-qscintilla.h b/libgui/src/m-editor/octave-qscintilla.h
--- a/libgui/src/m-editor/octave-qscintilla.h
+++ b/libgui/src/m-editor/octave-qscintilla.h
@@ -101,16 +101,18 @@ namespace octave
   protected:
 
     void focusInEvent (QFocusEvent *focusEvent);
 
     void show_replace_action_tooltip (void);
 
     void keyPressEvent (QKeyEvent *e);
 
+    void dragEnterEvent (QDragEnterEvent *e);
+
   private:
 
     void auto_close (int auto_endif, int l,
                      const QString& line, QString& first_word);
 
     QString m_word_at_cursor;
 
     QString m_selection;
